### PR TITLE
Fix GHA hygiene workflow

### PR DESCRIPTION
**Pull request by dra27**
I noticed when rebasing some older PRs that the `check-typo` job was failing _on the push_. Closer inspection showed that while it was using the correct range of commits to determine which files to check, for a force push it was using the _old_ branch ref not the new one, which was causing failures. The first commit here fixes that (and the commit message includes the somewhat dull technical detail).

While I'm here, I've also switched the rather hacky (and potentially unstable) use of a cookie file to signal the failure of the "Minimal compiler" step to use `::set-output` and record the success properly in a step output variable. It doesn't particularly shorten anything, but it does make it slightly less unclear.

### CODE REVIEW

File: tools/ci/actions/runner.sh
Comment (gasche) [inline tools/ci/actions/runner.sh:128]: I'm unsure what this means, my best guess would be "variable 1, or `-1` if absent", why do we need a default here?

File: tools/ci/actions/runner.sh
Comment (gasche) [inline tools/ci/actions/runner.sh:139]: Well this magic is somehow more magical than the touch-a-file protocol (it requires more expert knowledge of `bash` to be understood), but it also makes the code cleaner.

File: tools/ci/actions/deepen-fetch.sh
Comment (gasche) [inline tools/ci/actions/deepen-fetch.sh]: Maybe comment on when this can happen in practice? (It's a bit mysterious to me.)

**Review by gasche — APPROVED** [review]
This is good. I left a few comments that @dra27 could address or not before we merge -- @dra27, please let us know when you are done handling them.

File: tools/ci/actions/runner.sh
Comment (dra27) [reply tools/ci/actions/runner.sh:128]: You can't pass parameters to a `trap` function, so we have to work out a default.

`${parameter:-word}` is POSIX sh for a default value see [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02)). So this means "use the contents of `$1` unless its empty or unset in which case use `1`".

File: tools/ci/actions/runner.sh
Comment (dra27) [reply tools/ci/actions/runner.sh:139]: Indeed - having realised (while working on another workflow) that the file trick was not _necessary_, what bothers me is that we have other checks which check for the tree being clean, etc.

File: tools/ci/actions/deepen-fetch.sh
Comment (dra27) [reply tools/ci/actions/deepen-fetch.sh]: Oh yes, the commit message explains why but not the comment! I changed it to "(if it hasn't been by a previous step)" which I think is clearer?

**Comment by dra27** [comment]
Good to go for me, if you're happy with the amended comment!

File: tools/ci/actions/runner.sh
Comment (gasche) [reply tools/ci/actions/runner.sh:128]: An alternative would be to have a ReportBuildStatus with a proper parameter, and then parameter-less `ReportBuildSuccess` and `ReportBuildFailure` one-liners on top of it. But okay for the current code if we expect "real" shell programmers to be familiar with it.

Comment (dra27) [reply tools/ci/actions/runner.sh:128]: Who’re you accusing of being a “real” shell programmer 😉

**Pull request merged**
